File: easy/3740. Minimum Distance Between Three Equal Elements I/solution.cpp
class Solution {
public:
    int tempMINdistance(vector<int> nums, int target)
    {
        vector<int> temp;
        for(int i = 0; i < nums.size(); i++)
        {
            if(nums[i] == target)
                temp.push_back(i);
        }

        int ans = INT_MAX;
        for(int i = 0; i <= temp.size() - 3; i++)
        {
            int t = abs(temp[i] - temp[i+1]) 
                  + abs(temp[i+1] - temp[i+2]) 
                  + abs(temp[i+2] - temp[i]);

            ans = min(ans, t);
        }

        return ans;
    }

    int minimumDistance(vector<int>& nums) {
        map<int,int> freq;
        vector<int> ans;

        for(int i = 0; i < nums.size(); i++)
        {
            freq[nums[i]]++;
            if(freq[nums[i]] == 3)
                ans.push_back(tempMINdistance(nums, nums[i]));
        }

        if(ans.empty()) return -1;
        return *min_element(ans.begin(), ans.end());
    }
};

File: easy/3740. Minimum Distance Between Three Equal Elements I/solution.rs
impl Solution {
    fn calc(nums: &Vec<i32>, target: i32) -> i32 {
        let mut pos = Vec::new();

        for i in 0..nums.len() {
            if nums[i] == target {
                pos.push(i as i32);
            }
        }

        let mut res = i32::MAX;

        for i in 0..=pos.len() - 3 {
            let a = pos[i];
            let b = pos[i + 1];
            let c = pos[i + 2];

            let cur = (a - b).abs() + (b - c).abs() + (c - a).abs();
            res = res.min(cur);
        }

        res
    }

    pub fn minimum_distance(nums: Vec<i32>) -> i32 {
        use std::collections::HashMap;

        let mut freq = HashMap::new();
        let mut res = i32::MAX;

        for &x in &nums {
            let cnt = freq.entry(x).or_insert(0);
            *cnt += 1;

            if *cnt == 3 {
                let val = Self::calc(&nums, x);
                res = res.min(val);
            }
        }

        if res == i32::MAX { -1 } else { res }
    }
}